File: src/main/java/dev/lonami/uniffidl/UdlStructureViewElement.java
package dev.lonami.uniffidl;

import com.intellij.ide.projectView.PresentationData;
import com.intellij.ide.structureView.StructureViewTreeElement;
import com.intellij.ide.util.treeView.smartTree.SortableTreeElement;
import com.intellij.ide.util.treeView.smartTree.TreeElement;
import com.intellij.navigation.ItemPresentation;
import com.intellij.psi.NavigatablePsiElement;
import dev.lonami.uniffidl.psi.UdlDefinition;
import dev.lonami.uniffidl.psi.UdlFile;
import dev.lonami.uniffidl.psi.impl.UdlDefinitionImpl;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

public class UdlStructureViewElement implements StructureViewTreeElement, SortableTreeElement {
    private final NavigatablePsiElement myElement;

    public UdlStructureViewElement(NavigatablePsiElement element) {
        myElement = element;
    }

    @Override
    public Object getValue() {
        return myElement;
    }

    @Override
    public void navigate(boolean requestFocus) {
        myElement.navigate(requestFocus);
    }

    @Override
    public boolean canNavigate() {
        return myElement.canNavigate();
    }

    @Override
    public boolean canNavigateToSource() {
        return myElement.canNavigateToSource();
    }

    @Override
    public @NotNull String getAlphaSortKey() {
        String name = myElement.getName();
        return name != null ? name : "";
    }

    @Override
    public @NotNull ItemPresentation getPresentation() {
        ItemPresentation presentation = myElement.getPresentation();
        return presentation != null ? presentation : new PresentationData();
    }

    @Override
    public TreeElement @NotNull [] getChildren() {
        if (myElement instanceof UdlFile) {
            List<UdlDefinition> definitions = UdlUtil.findDefinitions((UdlFile) myElement);
            List<TreeElement> treeElements = new ArrayList<>(definitions.size());
            for (UdlDefinition definition : definitions) {
                treeElements.add(new UdlStructureViewElement((UdlDefinitionImpl) definition));
            }
            return treeElements.toArray(new TreeElement[0]);
        }
        return EMPTY_ARRAY;
    }
}

File: src/main/kotlin/dev/lonami/uniffidl/UdlStructureViewElement.kt
package dev.lonami.uniffidl

import com.intellij.ide.projectView.PresentationData
import com.intellij.ide.structureView.StructureViewTreeElement
import com.intellij.ide.util.treeView.smartTree.SortableTreeElement
import com.intellij.ide.util.treeView.smartTree.TreeElement
import com.intellij.navigation.ItemPresentation
import com.intellij.psi.NavigatablePsiElement
import dev.lonami.uniffidl.psi.UdlDefinition
import dev.lonami.uniffidl.psi.UdlFile
import dev.lonami.uniffidl.psi.impl.UdlDefinitionImpl

class UdlStructureViewElement(private val myElement: NavigatablePsiElement) :
    StructureViewTreeElement, SortableTreeElement {

    override fun getValue(): Any = myElement

    override fun navigate(requestFocus: Boolean) {
        myElement.navigate(requestFocus)
    }

    override fun canNavigate(): Boolean = myElement.canNavigate()

    override fun canNavigateToSource(): Boolean = myElement.canNavigateToSource()

    override fun getAlphaSortKey(): String {
        val name = myElement.name
        return name ?: ""
    }

    override fun getPresentation(): ItemPresentation {
        val presentation = myElement.presentation
        return presentation ?: PresentationData()
    }

    override fun getChildren(): Array<TreeElement> {
        return when (myElement) {
            is UdlFile -> {
                val definitions = UdlUtil.findDefinitions(myElement)
                val treeElements = mutableListOf<TreeElement>()
                for (definition in definitions) {
                    treeElements.add(UdlStructureViewElement(definition as UdlDefinitionImpl))
                }
                treeElements.toTypedArray()
            }
            is UdlDefinition -> {
                val children = UdlUtil.findDefinitionChildren(myElement)
                val treeElements = mutableListOf<TreeElement>()
                for (child in children) {
                    treeElements.add(UdlStructureViewElement(child))
                }
                treeElements.toTypedArray()
            }
            else -> EMPTY_ARRAY
        }
    }
}
